File: server/vesnai/okf/parse.py
"""Parse and serialize OKF concept documents.

A concept document is::

    ---
    <yaml frontmatter>
    ---
    <markdown body>

Serialization is deterministic and round-trip stable at the structural level:
``parse(dump(c)) == (c.frontmatter, c.body)`` for any concept produced by the
parser. Well-known keys are emitted first in OKF-recommended order for
readability; remaining keys keep their insertion order.
"""

from __future__ import annotations

import yaml

from vesnai.okf.model import RECOMMENDED_ORDER, Concept

_DELIM = "---"


class OKFParseError(ValueError):
    """Raised when a concept document cannot be parsed."""


def parse_concept(text: str) -> Concept:
    """Parse a concept document into a :class:`Concept`.

    Tolerant of a leading BOM and of CRLF line endings. The body is everything
    after the closing frontmatter delimiter, with a single leading newline
    stripped (so it survives a round-trip).
    """
    if text.startswith("\ufeff"):
        text = text[1:]
    normalized = text.replace("\r\n", "\n").replace("\r", "\n")

    if not normalized.startswith(_DELIM + "\n") and normalized.strip() != _DELIM:
        raise OKFParseError("document does not start with a '---' frontmatter delimiter")

    # Find the closing delimiter on its own line.
    lines = normalized.split("\n")
    # lines[0] == '---'
    closing_index: int | None = None
    for i in range(1, len(lines)):
        if lines[i] == _DELIM:
            closing_index = i
            break
    if closing_index is None:
        raise OKFParseError("frontmatter block is not closed with a '---' line")

    fm_text = "\n".join(lines[1:closing_index])
    body = "\n".join(lines[closing_index + 1 :])
    # A single newline immediately follows the closing delimiter in dump(); strip it.
    if body.startswith("\n"):
        body = body[1:]

    try:
        loaded = yaml.safe_load(fm_text) if fm_text.strip() else {}
    except yaml.YAMLError as exc:  # pragma: no cover - exercised via tests
        raise OKFParseError(f"invalid YAML frontmatter: {exc}") from exc

    if loaded is None:
        loaded = {}
    if not isinstance(loaded, dict):
        raise OKFParseError("frontmatter must be a YAML mapping")

    return Concept(frontmatter=dict(loaded), body=body)


def _ordered_frontmatter(frontmatter: dict) -> dict:
    ordered: dict = {}
    for key in RECOMMENDED_ORDER:
        if key in frontmatter:
            ordered[key] = frontmatter[key]
    for key, value in frontmatter.items():
        if key not in ordered:
            ordered[key] = value
    return ordered


def dump_concept(concept: Concept) -> str:
    """Serialize a :class:`Concept` to its document form (deterministic)."""
    ordered = _ordered_frontmatter(concept.frontmatter)
    fm_text = yaml.safe_dump(
        ordered,
        sort_keys=False,
        allow_unicode=True,
        default_flow_style=False,
    ).rstrip("\n")
    body = concept.body
    return f"{_DELIM}\n{fm_text}\n{_DELIM}\n{body}"

File: packages/okf_dart/lib/src/parse.dart
import 'package:yaml/yaml.dart';

import 'concept.dart';

const _delim = '---';

class OkfParseException implements Exception {
  final String message;
  OkfParseException(this.message);
  @override
  String toString() => 'OkfParseException: $message';
}

Concept parseConcept(String text) {
  var normalized = text;
  if (normalized.startsWith('\uFEFF')) {
    normalized = normalized.substring(1);
  }
  normalized = normalized.replaceAll('\r\n', '\n').replaceAll('\r', '\n');

  if (!normalized.startsWith('$_delim\n') && normalized.trim() != _delim) {
    throw OkfParseException("document does not start with a '---' delimiter");
  }

  final lines = normalized.split('\n');
  int? closing;
  for (var i = 1; i < lines.length; i++) {
    if (lines[i] == _delim) {
      closing = i;
      break;
    }
  }
  if (closing == null) {
    throw OkfParseException("frontmatter not closed with a '---' line");
  }

  final fmText = lines.sublist(1, closing).join('\n');
  var body = lines.sublist(closing + 1).join('\n');
  if (body.startsWith('\n')) body = body.substring(1);

  dynamic loaded;
  try {
    loaded = fmText.trim().isEmpty ? <String, dynamic>{} : loadYaml(fmText);
  } on YamlException catch (e) {
    throw OkfParseException('invalid YAML frontmatter: $e');
  }

  loaded ??= <String, dynamic>{};
  if (loaded is! Map) {
    throw OkfParseException('frontmatter must be a YAML mapping');
  }
  return Concept(frontmatter: _deepMap(loaded), body: body);
}

Map<String, dynamic> _deepMap(Map map) {
  final out = <String, dynamic>{};
  map.forEach((k, v) => out[k.toString()] = _deepValue(v));
  return out;
}

dynamic _deepValue(dynamic v) {
  if (v is Map) return _deepMap(v);
  if (v is List) return v.map(_deepValue).toList();
  return v;
}

String dumpConcept(Concept concept) {
  final ordered = <String, dynamic>{};
  for (final key in recommendedOrder) {
    if (concept.frontmatter.containsKey(key)) {
      ordered[key] = concept.frontmatter[key];
    }
  }
  concept.frontmatter.forEach((k, v) {
    if (!ordered.containsKey(k)) ordered[k] = v;
  });

  final fm = StringBuffer();
  _emitMap(fm, ordered, 0);
  final fmText = fm.toString().trimRight();
  return '$_delim\n$fmText\n$_delim\n${concept.body}';
}

void _emitMap(StringBuffer buf, Map<String, dynamic> map, int indent) {
  final pad = '  ' * indent;
  map.forEach((key, value) {
    if (value is Map) {
      buf.writeln('$pad$key:');
      _emitMap(buf, value.cast<String, dynamic>(), indent + 1);
    } else if (value is List) {
      if (value.isEmpty) {
        buf.writeln('$pad$key: []');
      } else {
        buf.writeln('$pad$key:');
        for (final item in value) {
          if (item is Map) {
            buf.writeln('$pad-');
            _emitMap(buf, item.cast<String, dynamic>(), indent + 2);
          } else {
            buf.writeln('$pad- ${_scalar(item)}');
          }
        }
      }
    } else {
      buf.writeln('$pad$key: ${_scalar(value)}');
    }
  });
}

String _scalar(dynamic value) {
  if (value == null) return 'null';
  if (value is bool) return value ? 'true' : 'false';
  if (value is num) return value.toString();
  final s = value.toString();
  // Quote when the scalar could be misread as YAML structure.
  final needsQuote = s.isEmpty ||
      RegExp(r'^[\s]|[\s]$').hasMatch(s) ||
      RegExp(r'[:#\[\]{}&*!|>%@`,]').hasMatch(s) ||
      RegExp(r'^(true|false|null|~|\d)').hasMatch(s);
  if (needsQuote) {
    return "'${s.replaceAll("'", "''")}'";
  }
  return s;
}
